README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ggthreed - 3d geoms and stats for ggplot2<img src="man/figures/logo-with-pause.gif" align="right"/>

[![Travis build
status](https://travis-ci.org/coolbutuseless/ggthreed.svg?branch=master)](https://travis-ci.org/coolbutuseless/ggthreed)
[![AppVeyor build
status](https://ci.appveyor.com/api/projects/status/github/coolbutuseless/ggthreed?branch=master&svg=true)](https://ci.appveyor.com/project/coolbutuseless/ggthreed)
![](https://img.shields.io/badge/lifecycle-alpha-orange.svg)

[`ggthreed`](https://github.com/coolbutuseless/ggthreed) is a collection
of [`ggplot2`](https://github.com/tidyverse/ggplot2) geoms which use the
[`threed`](https://github.com/coolbutuseless/threed) library.

What’s in the box:

  - `geom_threedpie()` for creating 3d pie charts.
      - **I am fully aware of the crimes against visualisation I am
        committing here.**
  - `stat_anaglyph()` for creating red-blue anaglyph images

## Installation

You can install from github

``` r
# install.packages("devtools")
devtools::install_github("coolbutuseless/threed")
devtools::install_github("coolbutuseless/ggthreed")
```

# Red/blue anaglyphs with `stat_anaglyph()`

**Grab your glasses and lower your expectations\!\!**

![](man/figures/glasses.jpg)

## Usage

  - Works with point, line and polygon geoms. Although “works” may be a
    strong term - the 3d effect is very weak even if you tweak all the
    parameters
  - It might work with other geoms, but no guarantees that it produces
    anything worth looking at.
  - parameters:
      - `zoffset`, `zscale` - for determining how `z` aesthetic
        influences position
      - `red`, `blue` - hex colours to use for red/blue
      - `switch` - switch position of the colours

## Example red/blue anaglyphs with points.

``` r
ggplot(mtcars) +
  geom_point(aes(mpg, y = wt, z = disp), stat = 'anaglyph', alpha = 0.5, zscale = 10) +
  theme_bw()
```

<img src="man/figures/README-unnamed-chunk-2-1.png" width="100%" />

### Example of combined 3d projection and anaglyph

``` r
library(threed)

camera_to_world <- look_at_matrix(eye = c(1.5, 1.75, 3), at = c(0, 0, 0))

obj <- threed::mesh3dobj$cube %>%
  transform_by(invert_matrix(camera_to_world)) %>%
  translate_by(c(0, 0, -3)) %>%
  perspective_projection()

ggplot(obj, aes(x, y, z = z, group = element_id)) +
  geom_polygon(fill = NA, colour='black', aes(size = hidden), stat = 'anaglyph',
               zscale = 0.05, zoffset = -1.4, zinvert = FALSE) +
  scale_linetype_manual(values = c('TRUE' = "FF", 'FALSE' = 'solid')) +
  scale_size_manual(values = c('TRUE' = 0.1, 'FALSE' = 0.5)) +
  theme_void() +
  theme(legend.position = 'none') +
  coord_equal()
```

<img src="man/figures/README-gallery-cube-1.png" width="100%" />

## Animated Anaglyph

### Animated Icosahedron

See `vignette('animated-anaglyph', package='ggthreed')`

![](vignettes/gif/animated-anaglyph.gif)

# 3d pie charts with `geom_threedpie()`

## Usage

  - Requires only an `x` variable - which must be discrete.
  - Default stat is `count`
  - Adjustable parameters:
      - camera position
      - pie height
      - starting angle for first pie slice
      - tilt angle of pie relative to camera

## Issues/Limitations

  - `geom_threedpie()` overrides the aspect ratio of the plot it is
    displaying. This is a gigantic hack\! It means it is not possible to
    use `coord_fixed()` etc to change the plot aspec ratio.
  - The initial panel grid and x- and y-axes are still generated by the
    geom. Currently the only way to turn this off it to use
    `theme_void()`. Not sure how to disable this from within the `geom`.
  - Pie slice resolution is in increments of 2 degrees.

## Example - Simple Pie Chart

``` r
ggplot(mtcars) + 
  geom_threedpie(aes(x = as.factor(cyl))) + 
  theme_void() + 
  theme(legend.position = 'bottom')
```

<img src="man/figures/README-example-simple-1.png" width="500" />

## Example - Facetted Pie Chart

``` r
ggplot(diamonds)  +
  geom_threedpie(aes(as.factor(cut))) +
  facet_wrap(~clarity, labeller = label_both) +
  labs(title = "Distribution of Diamond Cuts by Clarity") +
  scale_fill_brewer(name = "Cut", palette = 'Set2') +
  theme_void()
```

<img src="man/figures/README-example-facet-1.png" width="100%" />

## Example - Behind the Scenes

  - This plot shows the polygons making up the pie.
  - Triangular polygons are rendered on top, and quadrilaterals are
    rendered for the side.
  - The `threed` library is used to rotate the assembled polygons and
    perform perspective projection.
  - By plotting the polygons in order of distance from the camera
    (furtherest polygons first), polygons which are behind others are
    hidden from view.
  - The pie has no bottom.
  - The quadrilaterals making up the side are the same colour as the
    top, but darkened by 15%.

<!-- end list -->

``` r
ggplot(mtcars) + 
  geom_threedpie(aes(x = as.factor(cyl)), alpha = 0.5) + 
  theme_void() + 
  theme(legend.position = 'bottom')
```

<img src="man/figures/README-example-behind-1.png" width="100%" />

## Example - Pie configuration

User adjustable:

  - pie height
  - pie tilt
  - start of first pie slice

![](man/figures/pie-anim.gif)


NEWS.md

# ggthreed 0.1.3

* Change x position adjustment for `StatAnaglyph` to vary with 1/z

# ggthreed 0.1.2

* Add `StatAnaglyph` for creating red-blue anaglyphs

# ggthreed 0.1.1

* Fix issue #3
    * 'group' is now correctly aligned between the 3d object polygons
      and the 'data' preparsed by ggplot2.


# ggthreed 0.1.0

* 20181113 v0.1.0 initial release
* `geom_threedpie()` added


R/stat-anaglyph.R


'%||%' <- function(x, y) {
  if (is.null(x)) {
    y
  } else {
    x
  }
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Red-blue anaglyph
#'
#' Split data to create red-blue anaglyph
#'
#' @inheritParams ggplot2::layer
#' @inheritParams ggplot2::geom_point
#' @param zoffset,zscale how to offset the \code{x} coordinates based upon the
#' z Default zoffset = 0, zscale = 1
#' @param zinvert Does z get bigger as a point moves further away. default: TRUE
#' @param red,blue colours to use for the anaglyph
#' @param switch switch sides for red/blue?  default: FALSE
#'
#' @import ggplot2
#' @export
#'
#' @examples
#' \dontrun{
#' ggplot(mtcars) +
#'    geom_point(aes(wt, mpg, z = disp), stat = 'anaglyph') +
#'    theme_bw()
#' }
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
stat_anaglyph <- function(mapping     = NULL,
                          data        = NULL,
                          geom        = "point",
                          position    = "identity",
                          zoffset     = 0.0,
                          zscale      = 1.0,
                          zinvert     = TRUE,
                          red         = "#ff0000",
                          blue        = "#00fffb",
                          switch      = FALSE,
                          ...,
                          show.legend = NA,
                          inherit.aes = TRUE) {
  layer(
    data = data,
    mapping     = mapping,
    stat        = StatAnaglyph,
    geom        = geom,
    position    = position,
    show.legend = show.legend,
    inherit.aes = inherit.aes,
    params = list(
      zoffset = zoffset,
      zscale  = zscale,
      zinvert = zinvert,
      red     = red,
      blue    = blue,
      switch  = switch,
      na.rm   = TRUE,
      ...
    )
  )
}

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' StatAnaglyph
#'
#' @format NULL
#' @usage NULL
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
StatAnaglyph <- ggproto("StatAnaglyph", Stat,

  required_aes = c("x", "z"),
  default_aes = aes(
    zoffset = 0.0,
    zscale  = 1,
    zinvert = TRUE,
    switch  = FALSE,
    red     = "#ff0000",
    blue    = "#00fffb"
  ),

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Need to set arguments here so ggplot is aware of the params
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  compute_group = function(data, scales, zoffset, zscale, switch, red, blue, zinvert) {
    data
  },

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # All the work for this Stat happens in the finish_layer
  # i.e. all the stats have been adjusted etc
  # In here we duplicate the entire dataset, ensure the groups are disjoint
  # and the colours to red/blue in the 2 datasets
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  finish_layer = function(self, data, params) {

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Set a 'z' var if there isn't one
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    if (!exists('z', data)) {
      data$z <- 1
    }

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Set a 'group' var if there isn't one
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    if (!exists('group', data)) {
      data$group <- 1
    }

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Ensure groups start at 1, and then find the max_group
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    data      <- transform(data, group = group - min(group) + 1L)
    max_group <- max(data$group, na.rm = TRUE)

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Set default values for unset params
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    params$zoffset <- params$zoffset %||% self$default_aes$zoffset
    params$zscale  <- params$zscale  %||% self$default_aes$zscale
    params$zinvert <- params$zinvert %||% self$default_aes$zinvert
    params$switch  <- params$switch  %||% self$default_aes$switch
    params$red     <- params$red     %||% self$default_aes$red
    params$blue    <- params$blue    %||% self$default_aes$blue

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # If 'switch' is TRUE, swap the red and the blue
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    if (params$switch) {
      params$zscale <- params$zscale * -1
    }

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Calculate the adjustment of the x value depending on the z value
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    if (params$zinvert) {
      data <- transform(data, zadjust = (params$zscale / (z + params$zoffset)))
    } else {
      data <- transform(data, zadjust = (params$zscale * (z + params$zoffset)))
    }

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Transform the data into red/blue versions
    # Adjust the x coord of the points based upon the z
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    red_data  <- transform(data, x = x - zadjust, colour = params$red)
    blue_data <- transform(data, x = x + zadjust, colour = params$blue, group = group + max_group)


    rbind(red_data, blue_data)
  }
)



if (interactive()) {
  library(ggplot2)
  library(ggthreed)

  ggplot(mtcars) +
    geom_line(aes(mpg, y = wt, z = disp), stat = 'anaglyph', switch = TRUE,
               zscale = 10, alpha = 0.5) +
    theme_bw()

}




R/geom-threedpie.R


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Clone of ggplot2:::empty()
#'
#' @param df data.frame
#'
#' @return TRUE if empty (NULL, no rows or no cols)
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
empty <- function (df) {
  is.null(df) || nrow(df) == 0 || ncol(df) == 0
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Darken a hex colour by the given amount
#' Stolen from \url{https://benjaminlmoore.wordpress.com/2014/03/18/guardian-data-blog-uk-elections/}
#'
#' @param hex_colour strings e.g. "#345678"
#' @param amount fraction to darken by. default 0.15
#'
#' @return darkened hex colours
#'
#' @importFrom grDevices col2rgb rgb
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
darken_colours <- function(hex_colour, amount = 0.15) {
  if (amount < 0 || amount > 1) {
    stop("darken_colours(): amount must be beween 0 and 1.")
  }
  return(rgb(t(col2rgb(hex_colour) * (1 - amount)), maxColorValue = 255))
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' 3d pie charts in ggplot2
#'
#'
#' @inheritParams ggplot2::layer
#'
#' @param start_degrees starting angle for first pie slice (in degrees). Default: 0
#' @param tilt_degrees angle by which to tilt the pie towards the camera (in degrees). Default: 0.
#' @param height height of the pie. Default: 0.1
#' @param darken fraction by which to darken colours for the side of the pie. Default: 0.15
#' @param camera_eye location of camera eye. Default: c(0, 3, 5)
#' @param camera_look_at at what point is the camera looking. Default: c(0, 0, 0)
#' @param na.rm If `FALSE`, the default, missing values are removed with
#'   a warning. If `TRUE`, missing values are silently removed.
#' @param ... Other arguments passed on to [layer()]. These are
#'   often aesthetics, used to set an aesthetic to a fixed value, like
#'   `colour = "red"` or `size = 3`. They may also be parameters
#'   to the paired geom/stat.
#'
#' @examples
#' \dontrun{
#'   ggplot(mtcars)  +
#'     geom_threedpie(aes(as.factor(cyl))) +
#'     coord_equal() +
#'     theme_minimal()
#' }
#'
#' @importFrom threed fortify.mesh3d
#' @import ggplot2
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
geom_threedpie <- function(mapping        = NULL,
                           data           = NULL,
                           stat           = "count",
                           position       = "identity",
                           ...,
                           start_degrees  = 0,
                           tilt_degrees   = 0,
                           height         = 0.1,
                           darken         = 0.15,
                           camera_eye     = c(0, 3, 5),
                           camera_look_at = c(0, 0, 0),
                           na.rm          = FALSE,
                           show.legend    = NA,
                           inherit.aes    = TRUE) {

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Set fill to be the 'x' variable unless already set by user.
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  if (!"fill" %in% names(mapping)) { mapping$fill   <- mapping$x }

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Don't allow mapping colour. It is just mapped to be the same as the 'fill'
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  mapping$colour <- NULL

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Create the layer
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  layer(
    data        = data,
    mapping     = mapping,
    stat        = stat,
    geom        = GeomThreedPie,
    position    = position,
    show.legend = show.legend,
    inherit.aes = inherit.aes,
    params      = list(
      start_degrees  = start_degrees,
      tilt_degrees   = tilt_degrees,
      height         = height,
      darken         = darken,
      camera_eye     = camera_eye,
      camera_look_at = camera_look_at,
      na.rm          = na.rm,
      ...
    )
  )
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Create a new ggproto object
#'
#' @section Calling methods in a parent:
#' To explicitly call a methods in a parent, use
#' `ggproto_parent(Parent, self)`.
#'
#' @param _class Class name to assign to the object. This is stored as the class
#'   attribute of the object. This is optional: if `NULL` (the default),
#'   no class name will be added to the object.
#' @param _inherit ggproto object to inherit from. If `NULL`, don't
#'   inherit from any object.
#' @param ... A list of members in the ggproto object.
#'
#' @format NULL
#' @usage NULL
#'
#' @import grid
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
GeomThreedPie <- ggproto(
  "GeomThreedPie", Geom,


  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Default aesthetics if none given
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  default_aes = aes(
    colour   = NA,
    fill     = NA,
    size     = 0.5,
    linetype = 1,
    alpha    = 1
  ),


  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Function to handle NAs
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  handle_na = function(data, params) {
    data
  },


  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Which aesthetics are absolutely REQUIRED otherwise there's an error
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  required_aes = c('x'),


  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # How to draw a legend key
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  draw_key = draw_key_polygon,


  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Drawing a particular panel
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  draw_panel = function(data, panel_params, coord, start_degrees, tilt_degrees,
                        height, darken, camera_eye, camera_look_at) {

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Enforce 'x' to be a discrete variable. This hack was the best I could do.
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    if (all(data$group == -1)) {
      stop("geom_threedpie: 'x' must be discrete.")
    }

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Create the pie mesh3d objects
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    objs <- create_pie_objects(data$group, data$count,
                               start_degrees  = start_degrees,
                               tilt_degrees   = tilt_degrees,
                               height         = height,
                               camera_eye     = camera_eye,
                               camera_look_at = camera_look_at)

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Convert the mesh3d objects to data.frames
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    pie     <- as.data.frame(objs$pie)

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Align the groups correctly
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    pie$group <- data$group[pie$group]

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Drop the x + y from the prepared geomdata
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    summary_df <- data %>% select(-x, -y)

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # But attach the rest to the polygon data.
    # Set the grouping variable to be the 'zorder' of the polygon so that
    # they get drawn in order from back to front
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    polygon_df <- pie %>%
      left_join(summary_df, by = 'group') %>%
      as.tbl() %>%
      mutate(group = zorder)

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Darken the quads that make up the side of the pie
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    polygon_df <- polygon_df %>%
      mutate(
        fill   = if_else(element_type == 4, darken_colours(fill, amount = darken), fill),
        colour = fill
      )

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Add a bit of a buffer around the plotting range
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    panel_params$y.range <- extendrange(x = polygon_df$y)
    panel_params$x.range <- extendrange(x = polygon_df$x)

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Calculate the new plot ratio
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    new_ratio <- diff(panel_params$y.range) / diff(panel_params$x.range)

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Adjust the plot aspect ratio to match the pie chart dimensions
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    coord$aspect <- function(...) {new_ratio}

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Create the complete grobTree objects which make up the pie.
    # This was initially a more complicated collection of grobs.
    # Leaving in the grobTree()/nullGrob() calls for future expansion and a
    # template for other ggthreed geoms.
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    ggplot2:::ggname(
      "geom_threedpie",
      grobTree(
        GeomPolygon$draw_panel(polygon_df, panel_params, coord),
        grid::nullGrob()
      )
    )

  }  # end of draw_panel()
)




#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Testing
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
if (FALSE) {
  library(grid)
  library(ggplot2)
  library(threed)
  library(ggthreed)
  library(dplyr)

  ggplot(mtcars)  +
    geom_threedpie(aes(as.factor(cyl)), tilt_degrees = -20) +
    coord_equal() +
    theme_minimal() +
    # theme(legend.position = 'none') +
    NULL


  ggplot(diamonds)  +
    geom_threedpie(aes(as.factor(cut)), tilt_degrees = 0) +
    theme_minimal() +
    facet_wrap(~clarity, labeller = label_both) +
    labs(
      title = "Distribution of Diamond Cuts by Clarity"
    ) +
    scale_fill_brewer(name = "Cut", palette = 'Dark2') +
    NULL
}

